### ISSUE TITLE
distsql: the hashRouter should be zero-copy in common cases

### ISSUE THREAD

**Issue by andreimatei** [cockroachdb/cockroach#13829]
The `hashRouter` needs to hash rows consistently across nodes. It does this by re-encoding rows with a standard encoding (decided statically - currently `DatumEncoding_ASCENDING_KEY`), copying the recoded values to a byte buffer, and hashing the resulting byte stream.

The planning process should choose an encoding that minimizes the data to be recoded (so, the hash encoding should depend on the query), and also the `hashRouter` should not make copies of the data in case the desired encoding matches.

**Comment by andy-kimball** [comment]
This is about DistSQL planning rather than logical SQL planning. @jordanlewis, I think we need to keep the labels separate for those two cases until we (eventually) fold DistSQL planning into the cost-based optimizer.

**Comment by jordanlewis** [comment]
Agreed - this really should be `execution` labeled anyway. Not sure why I switched the labels before, makes more sense like this.

**Comment by aloknnikhil** [comment]
Anyone assigned to this?
